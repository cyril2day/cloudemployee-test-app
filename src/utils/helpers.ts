type ParamObject = {[name: string]: any }


export function normalize(parent: ParamObject) {
   if (parent && parent.children) {
      parent.children.forEach((child: any) => {
         if (!child.parentId)
            child.parentId = parent.id

         normalize(child)
      })
   }
}

export function flatten(items: ParamObject[]) {
   const temp: unknown[] = []

   items.forEach(item => {
      normalize(item)
      const { children, ...props } = item 

      temp.push(props)

      if (children && Array.isArray(children)) {
         temp.push(flatten(children))
      }
   })

   return temp.flat()
}
