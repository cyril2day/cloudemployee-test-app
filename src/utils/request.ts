import axios from 'axios'


const api = axios.create({ baseURL: 'https://cloudemployee-backend-test-app.vercel.app/api' })

api.interceptors.request.use(
   config => {
      return config
   }
)

api.interceptors.response.use(
   response => {
      if (response.data)
         return response
      else
         return Promise.reject(response)
   },
   error => Promise.reject(error.response)
)

export { api }
