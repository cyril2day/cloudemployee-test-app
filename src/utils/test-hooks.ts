import { MountingOptions } from "@vue/test-utils"
import { setActivePinia, createPinia } from 'pinia'

export type TestMountingOptions = MountingOptions<unknown>

export const StoreBeforeEach = () => {
   beforeEach(() => {
      setActivePinia(createPinia())
   })
}
