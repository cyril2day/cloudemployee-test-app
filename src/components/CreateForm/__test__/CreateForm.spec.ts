import { mount, VueWrapper } from "@vue/test-utils"
import { mergeWith } from "lodash"
import { TestMountingOptions } from "../../../utils/test-hooks"
import CreateFormVue from "../CreateForm.vue"


const createWrapper = (overrides?: TestMountingOptions) => {
   const defaultMountingOption: TestMountingOptions = {
      global: {
         //
      }
   }

   return mount(CreateFormVue, mergeWith(defaultMountingOption, overrides))
}

describe('Create Form Page Unit Test', () => {
   let wrapper: VueWrapper<any>

   beforeEach(() => {
      wrapper = createWrapper()   
   })

   test('Form header renders correct default title', () => {
      expect.assertions(1)
      const formTitleDefault = wrapper.find('.form-header')

      expect(formTitleDefault.html()).toContain('Create Sub Business Function')
   })

   test('input for item name is the fallback form content', () => {
      expect.assertions(1)
      const input = wrapper.findAll('input')

      expect(input.length).toBe(1)
   })

   test(`clicking submit will emit 'formSubmit' custom event together with the name and type as payload`, async() => {
      expect.assertions(2)
      const submit = wrapper.find('#formSubmit')

      expect(submit.element.nodeName).toBe('BUTTON')

      await submit.trigger('click')

      expect(wrapper.emitted('formSubmit')[0][0]).toEqual({ name: '', type: 'Sub Business Function' })
   })

   test('slot renders correctly', () => {
      expect.assertions(2)
      const wrapper = createWrapper({
         slots: {
            'form-header': '<h3>Create Activity</h3>',
            default: '<div class="sample">Sample Content</div>'
         }
      })

      const header = wrapper.find('.form-header').html()
      const custom = wrapper.find('.sample').html()

      expect(header).toContain('Create Activity')
      expect(custom).toContain('Sample Content')
   })
})


describe('Create Form Component Snapshot Test', () => {
   test('Elements and styles renders correctly', () => {
      expect.assertions(1)
      const wrapper = createWrapper()

      expect(wrapper.element).toMatchSnapshot()
   })
})
