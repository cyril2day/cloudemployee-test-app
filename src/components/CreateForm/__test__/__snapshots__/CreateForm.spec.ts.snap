// Vitest Snapshot v1

exports[`Create Form Component Snapshot Test > Elements and styles renders correctly 1`] = `
<div
  class="form__wrapper"
>
  <!-- Form Header -->
  <div
    class="form-header"
  >
    
    <h3>
      Create Sub Business Function
    </h3>
    
  </div>
  <!-- Form Contents -->
  <div
    class="form-item"
  >
    <label
      for="name"
    >
      Name
    </label>
    <input
      autocomplete="off"
      id="name"
      name="name"
      type="text"
    />
  </div>
  
  
  <!-- Form Actions -->
  
  <div
    class="form-item"
  >
    <button
      id="formSubmit"
    >
      SUBMIT
    </button>
  </div>
  
</div>
`;
